Plugins/Public/mobiledocking_plugin/Main.cpp

```diff
@@ -628,270 +972,182 @@ bool UserCmd_Process(uint client, const wstring &wscCmd)
+	{
+		idToDockedInfoMap[iClientID] = &nameToDockedInfoMap[charname];
+		uint carrierClientID = HkGetClientIdFromCharname(idToDockedInfoMap[iClientID]->carrierName.c_str());
+		if (carrierClientID != -1)
 		{
-			uint clientID;
-			try
+			const auto& shipInfo = Archetype::GetShip(Players[iClientID].iShipArchetype);
+			wstring& shipName = HkGetWStringFromIDS(shipInfo->iIdsName);
+			PrintUserCmdText(carrierClientID, L"INFO: Docked %ls, %ls is standing by for launch", shipName.c_str(), charname);
+
+			const auto& carrierInfo = Players[carrierClientID];
+			const auto& sysInfo = Universe::get_system(carrierInfo.iSystemID);
+			wstring& sysName = HkGetWStringFromIDS(sysInfo->strid_name);
+			if (carrierInfo.iShipID)
 			{
-				clientID = HkGetClientIdFromPD(pPD);
+				Vector pos;
+				Matrix ori;
+				pub::SpaceObj::GetLocation(carrierInfo.iShipID, pos, ori);
+				wstring& sector = VectorToSectorCoord(sysInfo->id, pos);
+				PrintUserCmdText(iClientID, L"Carrier in flight, %ls %ls", sysName.c_str(), sector.c_str());
 			}
-			catch (...)
+			else if (carrierInfo.iBaseID)
 			{
-				clientID = 0;
+				const auto& baseInfo = Universe::get_base(carrierInfo.iBaseID);
+				wstring& baseName = HkGetWStringFromIDS(baseInfo->iBaseIDS);
+				PrintUserCmdText(iClientID, L"Carrier currently docked on %ls, %ls", baseName.c_str(), sysName.c_str());
 			}
-
-			string ID;
-
-			if (clientID == 0)
-				ID = "<Error>";
 			else
-				ID = to_string(clientID);
-
-			wstring charname;
-			wstring wscIP;
-
-			try
-			{
-				charname = (const wchar_t*)Players.GetActiveCharacterName(clientID);
-			}
-			catch (...)
-			{
-				charname = L"<NotLogged>";
-			}
-
-			try
-			{
-				HkGetPlayerIP(clientID, wscIP);
-			}
-			catch (...)
-			{
-				wscIP = L"<Error>";
-			}
-
-			string Charname = wstos(charname);
-			string State = "";
-
-			if (clientID > MAX_CLIENT_ID)
-				State += "Out of range";
-
-
-			if (find(SERVERcharnames.begin(), SERVERcharnames.end(), charname) != SERVERcharnames.end())
-			{
-				if (State != "")
-					State += " | ";
-				State += "DoubledName";
-			}
-
-			if (find(IDs.begin(), IDs.end(), clientID) != IDs.end())
 			{
-				if (State != "")
-					State += " | ";
-				State += "DoubledID";
+				PrintUserCmdText(iClientID, L"Carrier online in %ls system. Ready to launch.", sysName.c_str());
 			}
+		}
+		else
+		{
+			wstring& lastBaseName = GetLastBaseName(iClientID);
 
-			if (State == "")
-				State = "Fine";
-
-			if (charname != L"<NotLogged>")
-				SERVERcharnames.push_back(charname);
-			IDs.push_back(clientID);
-
-			SERVERlines.push_back(ID + " " + Charname + " " + wstos(wscIP) + " " + State);
+			PrintUserCmdText(iClientID, L"Carrier currently offline. Last docked base: %ls", lastBaseName);
 		}
+	}
+}
+
+void __stdcall UseItemRequest(SSPUseItem const& p1, unsigned int iClientID)
+{
+	returncode = DEFAULT_RETURNCODE;
+	if (disableRegensOnDockAttempt && !dockingInProgress.count(iClientID))

```

Comment:
is this ! supposed to be here?

Reply:
That's the whole reason I hooked UseItemRequest?
If it's configured to do so, any ship currently undergoing the docking procedure has it's shields disabled, cannot use bots and batts, and moving aborts the sequence. It is to make 'bot/battery tanking' not an option during the process, making it much more high-risk in combat.

This comes in tandem with an official rule change explicitly stating that docking on a Docking Module does NOT constituted a pvp death, and facilities inside allow for a full refill (bots, batts, ammo + repairs)

Reply:
I think this is written not just confusingly but wrong - right now, if you set disableRegensOnDockAttempt to false, this code is going to block all regen usage regardless of docking status. What you actually want is:
```c++
void __stdcall UseItemRequest(SSPUseItem const& p1, unsigned int iClientID)
{
    returncode = DEFAULT_RETURNCODE;
    if (disableRegensOnDockAttempt && dockingInProgress.count(iClientID))
    {
        static uint rejectSound = CreateID("ui_select_reject");
        pub::Audio::PlaySoundEffect(iClientID, rejectSound);
        returncode = SKIPPLUGINS_NOFUNCTIONCALL;
        return;
    }
}
```